File: units_cpp/src/unitsTest/cpp/constants/test_mars_properties.cc
/*
 * test_mars_properties.cc
 *
 *  Created on: Dec 27, 2017
 *      Author: sagis
 */

#include <units/constants/mars_properties.h>
#include <units/constants/physical_constants.h>
#include <units/units_multiply.h>
#include <units/units_ratio.h>
#include <units/units_pow.h>
#include <units/units_sqrt.h>
#include <units/planetary_constants_test_suite.h>

using namespace units::mars_properties;
using namespace units;
using units::physical_constants::gravitationalConstant;

class TestMarsProperties : public PlanetaryConstantsTestSuite
{
protected:
	TestMarsProperties()
	: PlanetaryConstantsTestSuite{0.6} // %
	{
	}
};

TEST_F(TestMarsProperties, test_surface_gravity)
{
	auto calculatedSurfaceGravity =
			units_ratio<units::acceleration::MetersPerSquareSecond>(
					auto_units_multiply(gravitationalConstant, marsMass),
					auto_units_multiply(marsEquatorialRadius, marsEquatorialRadius));
	checkEqualValue(marsSurfaceGravity, calculatedSurfaceGravity);
}

TEST_F(TestMarsProperties, test_mass)
{
	auto calculatedMass = (M_PI * 4 / 3) *
			units_multiply<units::mass::Kilograms>(
					marsMeanDensity ,units_pow<3>(marsEquatorialRadius));
	checkEqualValue(marsMass, calculatedMass);
}

TEST_F(TestMarsProperties, test_escape_velocity)
{
	auto calculatedEscapeVelocity = units_sqrt<units::speed::MetersPerSecond> (
			2 * auto_units_ratio(
					auto_units_multiply(gravitationalConstant, marsMass),
					marsEquatorialRadius
				)
			);
	checkEqualValue(marsEscapeVelocity, calculatedEscapeVelocity);
}

File: units_java/src/test/java/com/units/tests/constants/TestMarsProperties.java
package com.units.tests.constants;

import static com.units.constants.PhysicalConstants.gravitationalConstant;
import static com.units.constants.MarsProperties.*;

import org.junit.Test;

import com.units.acceleration.MetersPerSquareSecond;
import com.units.mass.Kilograms;
import com.units.speed.MetersPerSecond;

public class TestMarsProperties extends PlanetaryConstantsTestSuite {
	
	public TestMarsProperties() {
		super(0.6); // %
	}
	
	@Test
	public void testSurfaceGravity() {
		MetersPerSquareSecond calculatedAcceleration =
				gravitationalConstant
				.multiply(marsMass)
				.divide(marsEquatorialRadius)
				.divide(marsEquatorialRadius)
				.castTo(MetersPerSquareSecond.class);
		checkAcceleration(marsSurfaceGravity, calculatedAcceleration);	
	}
	
	@Test
	public void testMass() {
		Kilograms calculatedMass =
				marsMeanDensity
				.multiply(marsEquatorialRadius.pow(3))
				.castTo(Kilograms.class)
				.multiplyByScalar(Math.PI * 4 / 3);
		checkMass(marsMass, calculatedMass);
	}
	
	@Test
	public void testEscapeVelocity() {
		MetersPerSecond calculatedEscapeVelocity =
				gravitationalConstant
				.multiply(marsMass)
				.divide(marsEquatorialRadius)
				.sqrt()
				.castTo(MetersPerSecond.class)
				.multiplyByScalar(Math.sqrt(2));
		checkVelocity(marsEscapeVelocity, calculatedEscapeVelocity);
	}

}
